src/core/net/mdns.cpp

```diff
@@ -0,0 +1,3887 @@
+#include "common/locator_getters.hpp"
+#include "common/log.hpp"
+#include "common/numeric_limits.hpp"
+#include "common/type_traits.hpp"
+#include "instance/instance.hpp"
+
+/**
+ * @file
+ *   This file implements the Multicast DNS (mDNS) per RFC 6762.
+ */
+
+namespace ot {
+namespace Dns {
+namespace Multicast {
+
+RegisterLogModule("MulticastDns");
+
+//---------------------------------------------------------------------------------------------------------------------
+// otPlatMdns callbacks
+
+extern "C" void otPlatMdnsHandleReceive(otInstance                  *aInstance,
+                                        otMessage                   *aMessage,
+                                        bool                         aIsUnicast,
+                                        const otPlatMdnsAddressInfo *aAddress)
+{
+    AsCoreType(aInstance).Get<Core>().HandleMessage(AsCoreType(aMessage), aIsUnicast, AsCoreType(aAddress));
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+// Core
+
+const char Core::kLocalDomain[]         = "local.";
+const char Core::kUdpServiceLabel[]     = "_udp";
+const char Core::kTcpServiceLabel[]     = "_tcp";
+const char Core::kSubServiceLabel[]     = "_sub";
+const char Core::kServciesDnssdLabels[] = "_services._dns-sd._udp";
+
+Core::Core(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mIsEnabled(false)
+    , mIsQuestionUnicastAllowed(kDefaultQuAllowed)
+    , mMaxMessageSize(kMaxMessageSize)
+    , mInfraIfIndex(0)
+    , mMultiPacketRxMessages(aInstance)
+    , mEntryTimer(aInstance)
+    , mEntryTask(aInstance)
+    , mTxMessageHistory(aInstance)
+    , mConflictCallback(nullptr)
+{
+}
+
+Error Core::SetEnabled(bool aEnable, uint32_t aInfraIfIndex)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(aEnable != mIsEnabled, error = kErrorAlready);
+    SuccessOrExit(error = otPlatMdnsSetListeningEnabled(&GetInstance(), aEnable, aInfraIfIndex));
+
+    mIsEnabled    = aEnable;
+    mInfraIfIndex = aInfraIfIndex;
+
+    if (mIsEnabled)
+    {
+        LogInfo("Enabling on infra-if-index %lu", ToUlong(mInfraIfIndex));
+    }
+    else
+    {
+        LogInfo("Disabling");
+    }
+
+    if (!mIsEnabled)
+    {
+        mHostEntries.Clear();
+        mServiceEntries.Clear();
+        mServiceTypes.Clear();
+        mMultiPacketRxMessages.Clear();
+        mTxMessageHistory.Clear();
+        mEntryTimer.Stop();
+    }
+
+exit:
+    return error;
+}
+
+template <typename EntryType, typename ItemInfo>
+Error Core::Register(const ItemInfo &aItemInfo, RequestId aRequestId, RegisterCallback aCallback)
+{
+    Error      error = kErrorNone;
+    EntryType *entry;
+
+    VerifyOrExit(mIsEnabled, error = kErrorInvalidState);
+
+    entry = GetList<EntryType>().FindMatching(aItemInfo);
+
+    if (entry == nullptr)
+    {
+        entry = EntryType::AllocateAndInit(GetInstance(), aItemInfo);
+        OT_ASSERT(entry != nullptr);
```

Comment:
Isn't it going to be a race-condition between the Probing (started in AllocateAndInit() -> Entry::Init()) and GetList<EntryType().Push(*entry)? Basically what if the probing timer expires before we push to the list?

Reply:
All OpenThread API calls and platform callbacks (including timer/alarm) should run from single OS context (OT stack is not re-entrant). 

> Basically what if the probing timer expires before we push to the list?

Even if timer expires, platform must ensure to signal alarm callback `otPlatAlarmMilliFired()` after current `Tasklet` and/or call to OT stack has returned.